### PR TITLE
🟡(change): cache organization response

### PR DESCRIPTION
**Pull request by ruancomelli**
Organization is pulled quite a bit from the application, but should not be updated often.

## Summary by Sourcery

This pull request introduces caching for organization data using Redis, removes the rainbow animation feature from the theme provider, and consolidates organization-related API endpoints and hooks. Additionally, it updates dependencies to include the new Redis client library.

- **New Features**:
    - Introduced caching for organization data using Redis to reduce database load and improve response times.
- **Enhancements**:
    - Removed the rainbow animation feature from the theme provider and related components.
    - Consolidated organization-related API endpoints and hooks for better consistency and maintainability.
- **Chores**:
    - Updated dependencies in go.mod to include new Redis client library.

### CODE REVIEW

**Comment by ruancomelli** [comment]
This is an experiment review for experiment `improve_review_guide`
Run ID: `improve_review_guide/run_2024-08-12T00-02-04_v1-21-0-79-g76ada53b5`
The benchmark review for this pull request can be found at https://github.com/sourcery-ai-experiments/Trenova/pull/19.
This pull request was cloned from `https://github.com/sourcery-ai-experiments/Trenova/pull/11`. _(Note: the URL is not a link to avoid triggering a notification on the original pull request.)_

<details>
<summary>Experiment configuration</summary>

```yaml
review_config:
  # User configuration for the review
  # - benchmark - use the user config from the benchmark reviews
  # - <value> - use the value directly
  user_review_config:
    enable_ai_review: true
    enable_rule_comments: false

    enable_complexity_comments: false
    enable_security_comments: false
    enable_tests_comments: false
    enable_comment_suggestions: false

    enable_claude_comments: true
    enable_functionality_review: false

    enable_pull_request_summary: false
    enable_review_guide: true

    enable_approvals: false

  ai_review_config:
    # The model responses to use for the experiment
    # - benchmark - use the model responses from the benchmark reviews
    # - llm - call the language model to generate responses
    model_responses:
      comments_model: benchmark
      comment_area_model: benchmark
      comment_validation_model: benchmark
      comment_suggestion_model: benchmark
      complexity_model: benchmark
      docstrings_model: benchmark
      functionality_model: benchmark
      overall_comments_model: benchmark
      security_model: benchmark
      tests_model: benchmark
      pull_request_summary_model: benchmark
      review_guide_model: llm

# The pull request dataset to run the experiment on
pull_request_dataset:
  # open-source PRs
- https://github.com/sourcery-ai-experiments/gpt-crawler/pull/1
- https://github.com/sourcery-ai-experiments/graphviz2drawio/pull/1
- https://github.com/sourcery-ai-experiments/graphviz2drawio/pull/2
- https://github.com/sourcery-ai-experiments/iptux/pull/5
- https://github.com/sourcery-ai-experiments/iptux/pull/6
- https://github.com/sourcery-ai-experiments/kalima/pull/3
- https://github.com/sourcery-ai-experiments/pydantic_tfl_api/pull/1
- https://github.com/sourcery-ai-experiments/pydantic_tfl_api/pull/2
- https://github.com/sourcery-ai-experiments/Trenova/pull/11
  # Sourcery PRs
- https://github.com/sourcery-ai/core/pull/4055
- https://github.com/sourcery-ai/core/pull/4263
- https://github.com/sourcery-ai/core/pull/4352
- https://github.com/sourcery-ai/core/pull/4364
- https://github.com/sourcery-ai/core/pull/4366
- https://github.com/sourcery-ai/core/pull/4390
- https://github.com/sourcery-ai/core/pull/4402
- https://github.com/sourcery-ai/core/pull/4445
- https://github.com/sourcery-ai/core/pull/4462
- https://github.com/sourcery-ai/core/pull/4463

review_comment_labels: []

# Benchmark reviews generated by running
#   python -m scripts.experiment benchmark <experiment_name>
benchmark_reviews:
- dataset_pull_request: https://github.com/sourcery-ai/core/pull/4463
  review_pull_request: https://github.com/sourcery-ai-experiments/core/pull/236
- dataset_pull_request: https://github.com/sourcery-ai-experiments/gpt-crawler/pull/1
  review_pull_request: https://github.com/sourcery-ai-experiments/gpt-crawler/pull/10
- dataset_pull_request: https://github.com/sourcery-ai-experiments/iptux/pull/5
  review_pull_request: https://github.com/sourcery-ai-experiments/iptux/pull/32
- dataset_pull_request: https://github.com/sourcery-ai-experiments/iptux/pull/6
  review_pull_request: https://github.com/sourcery-ai-experiments/iptux/pull/33
- dataset_pull_request: https://github.com/sourcery-ai-experiments/kalima/pull/3
  review_pull_request: https://github.com/sourcery-ai-experiments/kalima/pull/17
- dataset_pull_request: https://github.com/sourcery-ai/core/pull/4055
  review_pull_request: https://github.com/sourcery-ai-experiments/core/pull/258
- dataset_pull_request: https://github.com/sourcery-ai/core/pull/4263
  review_pull_request: https://github.com/sourcery-ai-experiments/core/pull/259
- dataset_pull_request: https://github.com/sourcery-ai/core/pull/4352
  review_pull_request: https://github.com/sourcery-ai-experiments/core/pull/260
- dataset_pull_request: https://github.com/sourcery-ai/core/pull/4364
  review_pull_request: https://github.com/sourcery-ai-experiments/core/pull/261
- dataset_pull_request: https://github.com/sourcery-ai/core/pull/4366
  review_pull_request: https://github.com/sourcery-ai-experiments/core/pull/262
- dataset_pull_request: https://github.com/sourcery-ai/core/pull/4390
  review_pull_request: https://github.com/sourcery-ai-experiments/core/pull/263
- dataset_pull_request: https://github.com/sourcery-ai/core/pull/4402
  review_pull_request: https://github.com/sourcery-ai-experiments/core/pull/264
- dataset_pull_request: https://github.com/sourcery-ai/core/pull/4445
  review_pull_request: https://github.com/sourcery-ai-experiments/core/pull/265
- dataset_pull_request: https://github.com/sourcery-ai/core/pull/4462
  review_pull_request: https://github.com/sourcery-ai-experiments/core/pull/266
- dataset_pull_request: https://github.com/sourcery-ai-experiments/graphviz2drawio/pull/1
  review_pull_request: https://github.com/sourcery-ai-experiments/graphviz2drawio/pull/23
- dataset_pull_request: https://github.com/sourcery-ai-experiments/graphviz2drawio/pull/2
  review_pull_request: https://github.com/sourcery-ai-experiments/graphviz2drawio/pull/24
- dataset_pull_request: https://github.com/sourcery-ai-experiments/pydantic_tfl_api/pull/1
  review_pull_request: https://github.com/sourcery-ai-experiments/pydantic_tfl_api/pull/19
- dataset_pull_request: https://github.com/sourcery-ai-experiments/pydantic_tfl_api/pull/2
  review_pull_request: https://github.com/sourcery-ai-experiments/pydantic_tfl_api/pull/20
- dataset_pull_request: https://github.com/sourcery-ai-experiments/Trenova/pull/11
  review_pull_request: https://github.com/sourcery-ai-experiments/Trenova/pull/19

```

</details>

**Comment by SourceryAI** [comment]
## Reviewer's Guide by Sourcery

This pull request introduces caching for organization data using Redis, removes the rainbow animation feature, and consolidates organization-related API endpoints and hooks. The changes focus on improving performance and simplifying the codebase.

Key implementation details:
1. Redis caching is implemented for organization data in the OrganizationService.
2. The rainbow animation feature and related components are removed from the theme provider.
3. Organization-related API endpoints are consolidated, with '/organizations/me' and '/organizations/details' merged into a single '/organizations/' endpoint.
4. The useUserOrganization and useOrganization hooks are combined into a single useOrganization hook.
5. Redis client initialization is added to the server setup.
6. Dependencies are updated to include the new Redis client library.

### File-Level Changes

| Files | Changes |
| ----- | ------- |
| `internal/api/services/organization.service.go`<br/>`pkg/models/organization.go`<br/>`pkg/redis/client.go` | Implemented Redis caching for organization data, including cache invalidation, and added necessary methods for Redis compatibility. |
| `web/frontend/src/components/ui/theme-provider.tsx`<br/>`web/frontend/src/components/layout/user-avatar-menu.tsx`<br/>`web/frontend/src/components/layout/topbar.tsx` | Removed the rainbow animation feature and related components from the theme provider and user interface. |
| `web/frontend/src/hooks/useQueries.ts`<br/>`web/frontend/src/services/OrganizationRequestService.ts`<br/>`internal/api/handlers/organization.handler.go` | Consolidated organization-related API endpoints and hooks, simplifying the frontend and backend interaction for organization data. |
| `internal/server/server.go`<br/>`cmd/serve.go`<br/>`config/server_config.go` | Added Redis client initialization to the server setup and configuration. |

---

<details>
<summary>Tips</summary>

- Trigger a new Sourcery review by commenting `@sourcery-ai review` on the pull request.
- Continue your discussion with Sourcery by replying directly to review comments.
- You can change your review settings at any time by accessing your [dashboard](https://app.sourcery.ai):
    - Enable or disable the Sourcery-generated pull request summary or reviewer's guide;
    - Change the review language;
- You can always [contact us](mailto:support@sourcery.ai) if you have any questions or feedback.

</details>

File: internal/server/server.go [hunk 1]
```diff
@@ -160,6 +166,23 @@ func (s *Server) InitDB() *bun.DB {
 	return s.DB
 }
 
+func (s *Server) InitCache() *redis.Client {
```
Comment (SourceryAI) [inline internal/server/server.go:169]: **suggestion:** Handle cache client initialization failure more gracefully in InitCache.

The current implementation logs a fatal error if the cache client fails to initialize. Consider handling this more gracefully, perhaps by falling back to a no-op cache client or retrying the initialization.

File: internal/api/services/organization.service.go [hunk 2]
```diff
@@ -40,30 +43,36 @@ func NewOrganizationService(s *server.Server) *OrganizationService {
...
 		logger:      s.Logger,
 		minio:       s.Minio,
+		cache:       s.Cache,
 		fileService: file.NewFileService(s.Logger, s.FileHandler),
 	}
 }
 
-func (s *OrganizationService) GetUserOrganization(ctx context.Context, buID, orgID uuid.UUID) (*models.Organization, error) {
-	org := new(models.Organization)
+func (s *OrganizationService) organizationCacheKey(orgID uuid.UUID) string {
```
Comment (SourceryAI) [inline internal/api/services/organization.service.go:51]: **suggestion (performance):** Review efficiency of MarshalBinary and UnmarshalBinary methods.

The `MarshalBinary` and `UnmarshalBinary` methods in `organization.go` use `sonic` for serialization. Ensure that this is the most efficient choice for your use case, considering both performance and resource usage.

**Review by SourceryAI — COMMENTED** [review]
Hey @ruancomelli - I've reviewed your changes - here's some feedback:

**Overall Comments**:
- Consider encapsulating repetitive error-handling patterns for Redis operations into utility functions to make the code cleaner.
- Ensure that the `InitCache` function in `server.go` handles the case where the Redis client fails to initialize to avoid potential runtime errors.
- Document the changes in API endpoints clearly, especially the renaming of `useUserOrganization` to `useOrganization` and the corresponding endpoint change from `/organizations/me` to `/organizations/`.
- Consider encapsulating repetitive error-handling patterns for Redis operations into utility functions to make the code cleaner.
- Ensure that the `InitCache` function in `server.go` handles the case where the Redis client fails to initialize to avoid potential runtime errors.
- Document the changes in API endpoints clearly, especially the renaming of `useUserOrganization` to `useOrganization` and the corresponding endpoint change from `/organizations/me` to `/organizations/`.

<details>
<summary>Here's what I looked at during the review</summary>

- 🟡 **General issues**: 2 issues found
- 🟢 **Documentation**: all looks good
</details>

***

[LangSmith trace](https://smith.langchain.com/o/7d8518ad-6dd3-5c7b-96d2-757769296335/projects/p/88560225-5072-48bc-967c-dc54da53f265/r/0f05b955-b5c8-4ec3-a588-2a608fd3b699?poll=true)

<sub>
Help me be more useful! Please click 👍 or 👎 on each comment to tell me if it was helpful.
</sub>

**Pull request closed without merging**